File: frontend/demo/component/grid/grid-range-selection.ts
import 'Frontend/demo/init'; // hidden-source-line
import '@vaadin/grid';
import '@vaadin/grid/vaadin-grid-selection-column.js';
import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import type { GridItemToggleEvent, GridSelectedItemsChangedEvent } from '@vaadin/grid';
import { getPeople } from 'Frontend/demo/domain/DataService';
import { applyTheme } from 'Frontend/demo/theme';
import type Person from 'Frontend/generated/com/vaadin/demo/domain/Person';

// tag::snippet[]
@customElement('grid-range-selection')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    applyTheme(root);
    return root;
  }

  @state()
  private items: Person[] = [];

  @state()
  private selectedItems: Person[] = [];

  private rangeStartItem?: Person;

  protected override async firstUpdated() {
    const { people } = await getPeople();
    this.items = people;
  }

  handleItemToggle(event: GridItemToggleEvent<Person>) {
    const { item, selected, shiftKey } = event.detail;

    // If the anchor point isn't set, set it to the current item
    this.rangeStartItem ??= item;

    if (shiftKey) {
      // Calculcate the range of items between the anchor point and
      // the current item
      const [rangeStart, rangeEnd] = [this.rangeStartItem, item]
        .map((i) => this.items.indexOf(i))
        .sort((a, b) => a - b);
      const rangeItems = this.items.slice(rangeStart, rangeEnd + 1);

      // Update the selection state of items within the range
      // based on the state of the current item
      const newSelectedItems = new Set(this.selectedItems);
      rangeItems.forEach((rangeItem) => {
        if (selected) {
          newSelectedItems.add(rangeItem);
        } else {
          newSelectedItems.delete(rangeItem);
        }
      });
      this.selectedItems = [...newSelectedItems];
    }

    // Update the anchor point to the current item
    this.rangeStartItem = item;
  }

  protected override render() {
    return html`
      <vaadin-grid
        .items="${this.items}"
        .selectedItems="${this.selectedItems}"
        @selected-items-changed="${(e: GridSelectedItemsChangedEvent<Person>) => {
          this.selectedItems = e.detail.value;
        }}"
        @item-toggle="${this.handleItemToggle}"
      >
        <vaadin-grid-selection-column></vaadin-grid-selection-column>
        <vaadin-grid-column path="firstName"></vaadin-grid-column>
        <vaadin-grid-column path="lastName"></vaadin-grid-column>
        <vaadin-grid-column path="email"></vaadin-grid-column>
      </vaadin-grid>
    `;
  }
}
// end::snippet[]

File: src/main/java/com/vaadin/demo/component/grid/GridRangeSelection.java
package com.vaadin.demo.component.grid;

import java.util.List;

import com.vaadin.demo.domain.Person;
import com.vaadin.flow.component.grid.Grid;
import com.vaadin.flow.component.grid.GridMultiSelectionModel;
import com.vaadin.flow.component.grid.dataview.GridListDataView;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.router.Route;
import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.demo.domain.DataService;

@Route("grid-range-selection")
public class GridRangeSelection extends Div {

    private Person rangeStartItem;

    public GridRangeSelection() {
        // tag::snippet[]
        Grid<Person> grid = new Grid<>(Person.class, false);
        grid.addColumn(Person::getFirstName).setHeader("First name");
        grid.addColumn(Person::getLastName).setHeader("Last name");
        grid.addColumn(Person::getEmail).setHeader("Email");

        List<Person> people = DataService.getPeople();
        grid.setItems(people);

        GridMultiSelectionModel<Person> selectionModel = (GridMultiSelectionModel<Person>) grid
                .setSelectionMode(Grid.SelectionMode.MULTI);

        selectionModel.addClientItemToggleListener(event -> {
            Person item = event.getItem();

            // If the anchor point isn't set, set it to the current item
            if (rangeStartItem == null) {
                rangeStartItem = item;
            }

            if (event.isShiftKey()) {
                // Calculcate the range of items between the anchor
                // point and the current item
                GridListDataView<Person> dataView = grid.getListDataView();
                int rangeStart = dataView.getItemIndex(rangeStartItem).get();
                int rangeEnd = dataView.getItemIndex(item).get();
                Person[] rangeItems = dataView.getItems()
                        .skip(Math.min(rangeStart, rangeEnd))
                        .limit(Math.abs(rangeStart - rangeEnd) + 1)
                        .toArray(Person[]::new);

                // Update the selection state of items within the range
                // based on the state of the current item
                if (event.isSelected()) {
                    selectionModel.selectItems(rangeItems);
                } else {
                    selectionModel.deselectItems(rangeItems);
                }
            }

            // Update the anchor point to the current item
            rangeStartItem = item;
        });
        // end::snippet[]

        add(grid);
    }

    public static class Exporter // hidden-source-line
            extends DemoExporter<GridRangeSelection> { // hidden-source-line
    } // hidden-source-line
}
